source-git-monorepos/README.md

```diff
@@ -50,21 +89,367 @@ packages:
+
+Check if the package names (`downstream_package_name`) are correct and indeed
+exist in the distribution. See [packit/packit#88].
+
+### packit prepare-sources
+
+The command should be updated similar to `packit srpm`, so that it's possible
+to limit changes only to select packages.
+
+## Impact on Packit API
+
+### Internal Design
+
+With monorepos needs to be taught to map a single upstream repository
+(`PackitAPI.up`) to multiple dist-git/downstream repositories. This means that
+`PackitAPI.dg` should become a dictionary holding the objects for each
+dist-git repository.
+
+In order to make the transition easier, we could keep `.dg` as it is, and make
+it to return the dist-git repo for the first (and only) package. If there are
+more packages defined, an exception should be raised. This way parts of the
+API and CLI which were not yet updated to support monorepos, could continue to
+work, but would fail if used with a monorepo configuration specifying multiple
+packages.
+
+### Inheritance (Overriding) of Configuration Keys
+
+One can override configuration defined on the top-level of `.packit.yaml` in a
+package-object. That is, all top-level values are inherited by package-objects.
+
+Configuration values can be further overridden in job objects, but only for
+the packages the job references (if any.)
+
+```mermaid
+flowchart BT
+    C(Job) --> B(Package)
+    B --> A(Top-level)
+```
+
+A (not very meaningful) example:
+
+```yaml
+srpm_build_deps:
+  - python3-pip
+  - python3-setuptools
+packages:
+  subpackage1:
+    package_paths:
+      - package1
+    srpm_build_deps:
+      - python3-pip
+      - python3-setuptools
+      - python3-packaging
+  subpackage2:
+    package_paths:
+      - package2
+jobs:
+  - job: copr_build
+    trigger: pull_request
+    srpm_build_deps:
+      - python3-pip
+      - python3-setuptools
+      - cowsay
+    targets:
+      - fedora-all
+  - job: copr_build
+    trigger: release
+    packages:
+      - subpackage1
+    targets:
+      - fedora-all
+      - epel-9
+```
+
+In the example above when building PRs in Copr, the `srpm_build_deps` defined
+in the job would be used for both packages.
+
+On the other hand, when building releases in Copr, only `subpackage1` is going
+to be built, and the `srpm_build_deps` will be the ones defined for the
+package in the `packages` section.
+
+The example above could be made more useful if extending lists would be
+supported by Packit. (YAML doesn't support this, one gets nested lists when
+using references in lists.)
+
+### Impact on Actions
+
+Because actions can now be defined on the top-level, package-level and
+job-level, one might ask, in what path these should be executed.
+
+Because a package can map to multiple sub-paths in the monorepo, the most
+sensible thing to do is to keep it simple and execute each action at the
+top-level, even if the action is defined on a job- or package-level.
```

Comment:
Yes, that was my concern in the first question. But if we preserve current use-cases and document the behaviour of monorepo users, I don't see a problem with choosing any approach.

Reply:
:+1: Exactly.